### PR TITLE
[更新完毕] scalar tensor

### PR DESCRIPTION
**Pull request by isdanni**
Related issue: #83

### CODE REVIEW

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 1]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
 msgstr "（可选，标量或tuple）张量的形状"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po:40]: （可选，标量或元组）张量的形状

Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "（可选，标量或元组）张量的形状"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 2]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
...
 
 #: ../../scalar_tensor.rst:19
 msgid "x = ti.var(ti.i32, shape=4)"
-msgstr ""
+msgstr "x = ti.var(ti.i32, shape=4)"
 
 #: ../../scalar_tensor.rst:21
 msgid "This creates a 4x3 *dense* tensor with ``float32`` elements: ::"
-msgstr "下列代码将创建由 ``float32`` 组成的4x3密集张量： ::"
+msgstr "这将创建一个带有 ``float32`` 元素的4x3 *dense* 张量： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "这将创建一个元素为 ``float32`` 类型的4x3 *稠密* 张量： ::"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Which one do you prefer？

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 3]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
...
-msgstr ""
+msgstr "x = ti.var(ti.f32, shape=(4, 3))"
 
 #: ../../scalar_tensor.rst:26
 msgid ""
-"If shape is ``()`` (empty tuple), then a 0-D tensor (scalar) is created: "
-"::"
-msgstr "如果形状设为 ``()`` （空元组），那将创建一个0维张量（即标量）： ::"
+"If shape is ``()`` (empty tuple), then a 0-D tensor (scalar) is created: ::"
+msgstr "如果shape是 ``()`` （空元组），则创建一个0-D张量（标量）： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "如果 shape 是 ``()`` （空元组），则创建一个0-D张量（标量）： ::"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 4]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
...
 
 #: ../../scalar_tensor.rst:29
 msgid "x = ti.var(ti.f32, shape=())"
-msgstr ""
+msgstr "x = ti.var(ti.f32, shape=())"
 
 #: ../../scalar_tensor.rst:31
 msgid "Then access it by passing ``None`` as index: ::"
-msgstr "然后以 ``None`` 为指标来访问它： ::"
+msgstr "然后通过传递 ``None`` 作为索引来访问它： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "随后通过传递 ``None`` 作为索引来访问它： ::"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Maybe this word is more suitable？What's your opinion🧐

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 5]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
...
 
 #: ../../scalar_tensor.rst:36
 msgid ""
-"If shape is **not provided** or ``None``, the user must manually "
-"``place`` it afterwards: ::"
-msgstr "如果 **不提供** 形状参数 ``shape`` ，那么稍后用户必须手动放置（place）那个张量： ::"
+"If shape is **not provided** or ``None``, the user must manually ``place`` it "
+"afterwards: ::"
+msgstr ""
+"如果形状是 **未提供** 或 **没有**，则用户必须在其后手动 ``放置`` 形状： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"如果形状是 **not provided（未提供）** 或 **None（空）**，则用户必须在其后手动 ``place（放置）`` 形状： ::"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Perhaps the reader will prefer a translation that has some connection with the original text?

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 6]
```diff
@@ -39,51 +40,49 @@ msgid "(optional, scalar or tuple) the shape of tensor"
...
-msgstr "（可选，标量或tuple）张量的形状"
+msgstr "（可选，标量或元组）请参见 :ref:`offset`"
 
 #: ../../scalar_tensor.rst:16
 msgid ""
-"For example, this creates a *dense* tensor with four ``int32`` as "
-"elements: ::"
-msgstr "例如，下列代码将创建由4个 ``int32`` 组成的密集张量： ::"
+"For example, this creates a *dense* tensor with four ``int32`` as elements: ::"
+msgstr "例如，这将创建一个具有四个 ``int32`` 作为元素的 *dense* 张量： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "例如，这将创建一个具有四个 ``int32`` 作为元素的 *稠密(dense)* 张量： ::"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: The word "dense" can be translated into "稠密" according to Wiki Suggestions

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 7]
```diff
@@ -93,19 +92,20 @@ msgid ""
...
-#, fuzzy
 msgid ""
-"Not providing ``shape`` allows you to *place* the tensor in a layout "
-"other than the default *dense*, see :ref:`layout` for more details."
-msgstr "不提供形状 ``shape`` 还可以使你定义张量为 *稀疏* 张量，参考见 :ref:`layout` 。"
+"Not providing ``shape`` allows you to *place* the tensor in a layout other than "
+"the default *dense*, see :ref:`layout` for more details."
+msgstr ""
+"不提供 ``shape`` 允许你将张量 *放置* 在默认的 *dense* 之外的布局中，有关更多详细"
+"信息，请参见： :ref:`layout` 。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"通过不指定参数 ``shape`` 的写法，你可以将张量 *放置* 在默认的 *稠密数据布局* 之外的其他布局中，有关更多详细"
"信息，请参见： :ref:`layout` 。"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: This sentence reads a little strange according to the word-for-word translation, maybe we can make some artistic changes here🧐

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 8]
```diff
@@ -151,73 +150,76 @@ msgstr ""
 
 #: ../../scalar_tensor.rst:77
 msgid "Accessing components"
-msgstr ""
+msgstr "访问组件"
 
 #: ../../scalar_tensor.rst:79
 msgid "You can access an element of the Taichi tensor by an index or indices."
-msgstr ""
+msgstr "您可以通过一个或多个索引来访问Taichi张量的元素。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "您可以通过一个或多个索引来访问 Taichi 张量的元素。"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 9]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
 
 #: ../../scalar_tensor.rst:83
-#, fuzzy
 msgid "(Tensor) the tensor of scalars"
-msgstr "标量组成的张量"
+msgstr "（张量）标量的张量"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "（张量）标量张量"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 10]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
+msgstr "b[2] = 5"
 
 #: ../../scalar_tensor.rst:100
 msgid ""
-"In Python, x[(exp1, exp2, ..., expN)] is equivalent to x[exp1, exp2, ...,"
-" expN]; the latter is just syntactic sugar for the former."
+"In Python, x[(exp1, exp2, ..., expN)] is equivalent to x[exp1, exp2, ..., "
+"expN]; the latter is just syntactic sugar for the former."
 msgstr ""
+"在Python中， x[(exp1, exp2, ..., expN)] 等效于 x[(exp1, exp2, ..., expN)] ; 后者"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"在 Python 中， x[(exp1, exp2, ..., expN)] 等效于 x[(exp1, exp2, ..., expN)] ; 后者"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Leave some space between English and Chinese to breathe👻

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 11]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
 
 #: ../../scalar_tensor.rst:84
 msgid "(scalar) index of the first tensor dimension"
-msgstr ""
+msgstr "第一个张量维度的（标量）索引"
 
 #: ../../scalar_tensor.rst:85
 msgid "(scalar) index of the second tensor dimension"
-msgstr ""
+msgstr "第二张量维度的（标量）索引"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "第二个张量维度的（标量）索引"
```

Comment (ArkhamWJZ) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Keep consistent ;)

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 12]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
 
 #: ../../scalar_tensor.rst:86
 msgid "(scalar) the element at ``[p, q, ...]``"
-msgstr ""
+msgstr "（标量） ``[p, q, ...]`` 处的元素"
 
 #: ../../scalar_tensor.rst:88
 msgid "This extracts the element value at index ``[3, 4]`` of tensor ``a``: ::"
-msgstr ""
+msgstr "这将提取张量 ``a`` 的索引在 ``[3, 4]`` 处的元素值： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "这将提取张量 ``a`` 在索引为 ``[3, 4]`` 处的元素值： ::"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 13]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
 
 #: ../../scalar_tensor.rst:91
 msgid "x = a[3, 4]"
-msgstr ""
+msgstr "x = a[3, 4]"
 
 #: ../../scalar_tensor.rst:93
 msgid "This sets the element value at index ``2`` of 1D tensor ``b`` to ``5``: ::"
-msgstr ""
+msgstr "这会将一维张量 ``b`` 的索引 ``2`` 处的元素值设置为 ``5``： ::"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "这会将一维张量 ``b`` 在索引 ``2`` 处的元素值设置为 ``5``： ::"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 14]
```diff
@@ -151,73 +150,76 @@ msgstr ""
...
 
 #: ../../scalar_tensor.rst:104
 msgid ""
-"The returned value can also be ``Vector`` / ``Matrix`` if ``a`` is a "
-"tensor of vector / matrix, see :ref:`vector` for more details."
+"The returned value can also be ``Vector`` / ``Matrix`` if ``a`` is a tensor of "
+"vector / matrix, see :ref:`vector` for more details."
 msgstr ""
+"如果 ``a`` 是 ``Vector`` / ``Matrix`` 的张量，则返回的值也可以是Vector / "
+"Matrix，有关更多详细信息，请参见： :ref:`vector`。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"如果 ``a`` 是由 ``Vector`` / ``Matrix`` 构成的张量，则返回的值也可以是 Vector / "
"Matrix，有关更多详细信息，请参见： :ref:`vector`。"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 15]
```diff
@@ -247,26 +256,37 @@ msgid ""
...
-"``n=2`` grandparent, etc."
+"(optional, scalar) the number of parent steps, i.e. ``n=1`` for parent, ``n=2`` "
+"grandparent, etc."
 msgstr ""
+"（可选，标量）父级步骤数，即父级为 ``n = 1``，祖父母级为 ``n = 2``，等等。"
 
 #: ../../scalar_tensor.rst:159
 msgid "(SNode) the parent of ``a``'s containing SNode"
-msgstr ""
+msgstr "（SNode）包含 ``a`` SNode的父级"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "（结构节点） ``a`` 所属结构节点的父类节点"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 16]
```diff
@@ -247,26 +256,37 @@ msgid ""
...
+"x.data_type()  # ti.i32"
 
 #: ../../scalar_tensor.rst:158
 msgid ""
-"(optional, scalar) the number of parent steps, i.e. ``n=1`` for parent, "
-"``n=2`` grandparent, etc."
+"(optional, scalar) the number of parent steps, i.e. ``n=1`` for parent, ``n=2`` "
+"grandparent, etc."
 msgstr ""
+"（可选，标量）父级步骤数，即父级为 ``n = 1``，祖父母级为 ``n = 2``，等等。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"（可选，标量）父级步数，即父级节点为 ``n = 1``，祖父级节点为 ``n = 2``，等等。"
```

**Review by ArkhamWJZ — CHANGES_REQUESTED** [review]
LGTM !🥳
Most of the changes are personal. Please feel free to update 🌈

**Comment by isdanni** [comment]
> LGTM !
> Most of the changes are personal. Please feel free to update

Thanks for the thorough review :rofl: Updated.

**Comment by ArkhamWJZ** [comment]
Thank you for your understanding. BTW, nice work🔥 !

**Review by ArkhamWJZ — APPROVED** [review]

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 17]
```diff
@@ -28,62 +29,60 @@ msgstr "定义"
...
 
 #: ../../scalar_tensor.rst:36
 msgid ""
-"If shape is **not provided** or ``None``, the user must manually "
-"``place`` it afterwards: ::"
-msgstr "如果 **不提供** 形状参数 ``shape`` ，那么稍后用户必须手动放置（place）那个张量： ::"
+"If shape is **not provided** or ``None``, the user must manually ``place`` it "
+"afterwards: ::"
+msgstr ""
+"如果形状是 **not provided（未提供）** 或 **None（空）**，则用户必须在其后手动 ``place（放置）`` 形状： ::"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: Reversed?

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 18]
```diff
@@ -151,73 +150,76 @@ msgstr ""
 
 #: ../../scalar_tensor.rst:77
 msgid "Accessing components"
-msgstr ""
+msgstr "访问组件"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: 访问分量

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 19]
```diff
@@ -28,62 +29,60 @@ msgstr "定义"
 
 #: ../../scalar_tensor.rst
 msgid "Parameters"
-msgstr ""
+msgstr "参量"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: 参数

**Review by archibate — COMMENTED** [review]
Thank for updating! I have some nits.

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po (hunk 19 above)
Comment (isdanni) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "参数"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po (hunk 18 above)
Comment (isdanni) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
msgstr "访问分量"
```

**Review by isdanni — COMMENTED** [review]
Apply changes

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 20]
```diff
@@ -93,19 +92,20 @@ msgid ""
...
-#, fuzzy
 msgid ""
-"Not providing ``shape`` allows you to *place* the tensor in a layout "
-"other than the default *dense*, see :ref:`layout` for more details."
-msgstr "不提供形状 ``shape`` 还可以使你定义张量为 *稀疏* 张量，参考见 :ref:`layout` 。"
+"Not providing ``shape`` allows you to *place* the tensor in a layout other than "
+"the default *dense*, see :ref:`layout` for more details."
+msgstr ""
+"通过不指定参数 ``shape`` 的写法，你可以将张量 *放置* 在默认的 *稠密数据布局* 之外的其他布局中，有关更多详细"
+"信息，请参见： :ref:`layout` 。"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"通过不指定参数 ``shape`` 的写法，你可以将张量 *放置* 为默认的 *稠密数据布局* 之外的其他布局中，有关更多详细"
"信息，请参见： :ref:`layout` 。"
```

File: locales/zh_CN/LC_MESSAGES/scalar_tensor.po [hunk 21]
```diff
@@ -28,62 +29,60 @@ msgstr "定义"
...
 
 #: ../../scalar_tensor.rst:36
 msgid ""
-"If shape is **not provided** or ``None``, the user must manually "
-"``place`` it afterwards: ::"
-msgstr "如果 **不提供** 形状参数 ``shape`` ，那么稍后用户必须手动放置（place）那个张量： ::"
+"If shape is **not provided** or ``None``, the user must manually ``place`` it "
+"afterwards: ::"
+msgstr ""
+"如果形状是 **not provided（未提供）** 或 **None（空）**，则用户必须在其后手动 ``place（放置）`` 形状： ::"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/scalar_tensor.po]:
```suggestion
"如果形状参数 **未提供** 或指定为 ``None``，则其后用户必须在手动放置 (place) 它： ::"
```
We don't translate python keywords like `None`.

Comment (isdanni) [reply locales/zh_CN/LC_MESSAGES/scalar_tensor.po]: got it thx :rofl: didn't get ur point the first time

**Review by archibate — APPROVED** [review]
Thank for translating and reviewing :) Everything looks so great now!

**Pull request merged**